### ISSUE TITLE
Add register buffer

### ISSUE THREAD

**Issue by ginty** [Origen-SDK/origen#25]
Many registers transaction protocols support a burst mode where writes to adjacent addresses can be auto incremented.

The goal here is to have a system that leverages this without the user having to think about it. A register write buffer should be added to Origen core to automatically stage writes so that adjacent updates can be queued and then presented as a collection to the protocol driver for dispatch. Individual drivers can then decide where to burst or dispatch them individually.

This system should have the following features:

* Ability to turn the buffer on/off at application level (default to off until Origen v1 and it is proven out)
* The system should only be enabled when it is turned on and a driver that will accept a collection of registers is present (this will be the default for new drivers, but existing drivers may not work when presented with a collection).

The buffer should be flushed in the following circumstances:

* A write to a non-adjacent register
* Any non-register-based operation that generates a tester cycle or adds microcode
* An attempt is made to alter the state of a register already in the buffer
* At the end of the pattern

**Comment by chrisnappi** [comment]
To start, I think the passing of multi-register objects down to the protocol would be fairly directed - using a buffer scheme that would automatically optimize is a cool idea longer-term though.  What I am suggesting is that at the app level we create a collection of registers that has a 'write' method.

If we do this, then really we would just need to update existing drivers to take the array and decompose it into multiple writes.  We don't really have a ton of protocols right now, so I don't think that would be all that much work.  And if we start by only support direct app-level collections of registers then it should prevent breakage that could be caused by an intelligent scheme.

**Comment by ginty** [comment]
Oh I see, yes that would be a good starting point to get the driver side in order and then we can move to Origen automatically identifying such collections on the fly in future.
I'm wondering if we should make it work similar to the pin group API, where you can define a register group and the wrapper then allows you to treat it like a big reg.

~~~ruby
# This API isn't supported yet either, but thinking we should add this as a shorthand way to define
# registers where all bits are RW but not explicitly named
reg :r1, 0
reg :r2, 4

reg_group :big_reg, :r1, :r2

r1.size      # => 32
r1.address # => 0 
r2.size       # => 32
r2.address # => 4

big_reg.size       # => 64
big_reg.address  # => 0
big_reg.write(0x2222_1111)
r1.data    # => 0x1111
r2.data    # => 0x2222

# The group would support this API for easy processing by the drivers:
big_reg.each do |reg|
  # Process child reg as normal
end
~~~

**Comment by chrisnappi** [comment]
Yep - that would work really well I think.  Some protocols like AHB actually support both autoincrement and 4/8 work lines, so this fits in nicely.

**Comment by ginty** [comment]
Cool, you doing it?

**Comment by chrisnappi** [comment]
I'll at least proto-type it - but will need peer review :D

**Comment by ginty** [comment]
Nice one, I guess you can either do it on a branch in the core repo, or on your own fork

**Comment by ginty** [comment]
Actually, should probably do it on a branch in the core repo, means that we can both commit to it directly which will be more efficient and is a benefit of being in the core team

**Comment by chrisnappi** [comment]
Okay - I am going to work on this in conjunction with the AHB protocol and testport driver - which will hopefully have basic operation working by the end of the week.  AHB protocol will be Origen.
